Bracket

JorgeCastilloPrz opened the pull request:
fixes #1023

* [x] `Bracket` 
* [x] `BracketLaws`
* [x] Instances for `Kleisli`, `DeferredK`, `FluxK`, `MonoK`, `FlowableK`, `ObservableK`, `MaybeK`, `SingleK`
* [x] `IOBracket` instance (this one is quite more complex)
* [x] Law tests passing for all the data types providing a Bracket instance.
* [x] IOConnection
* [x] IOConnectionTests

JorgeCastilloPrz commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOPlatform.kt:
```diff
@@ -0,0 +1,16 @@
+   * On top of the JVM this function uses Throwable#addSuppressed, available since Java 7. On top of JavaScript the
+   * function would return a CompositeException.
+   */
+  fun composeErrors(first: Throwable, rest: List<Throwable>): Throwable {
```
Used from `CancelUtils` which is used from `IOConnection`. Not sure if we want to get into this rabbit hole though.

JorgeCastilloPrz replied:
Used from `CancelUtils` which is used from `IOConnection`. Not sure if we want to get into this rabbit hole though.

pakoito commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/BracketLaws.kt:
```diff
@@ -0,0 +1,92 @@
+
+object BracketLaws {
+  inline fun <F> laws(BF: Bracket<F, Throwable>,
+                      noinline cf: (Int) -> Kind<F, Int>,
```
You don't need this parameter, as it's `just` from `Applicative`.

JorgeCastilloPrz replied:
Alright, removed now!

pakoito commented on modules/effects/arrow-effects-reactor/src/main/kotlin/arrow/effects/FluxK.kt:
```diff
@@ -27,6 +32,14 @@ data class FluxK<A>(val flux: Flux<A>) : FluxKOf<A>, FluxKKindedJ<A> {
+  fun <B> bracketCase(use: (A) -> FluxKOf<B>, release: (A, ExitCase<Throwable>) -> FluxKOf<Unit>): FluxK<B> =
+    flatMap { a ->
+      use(a).fix().flux
+        .doOnNext { release(a, ExitCase.Completed) }
```
It shouldn't be `doOnNext` because `a` can be used multiple times. It should be on `doOnComplete`.

In any case, check if `doOnTerminate` exists for both cases, and whether they're called when the flux is `cancelled`.

JorgeCastilloPrz replied:
I was mistaken about how the callbacks work, sorry. I'm refactored it to use the 3 different callbacks depending on the case now:

```
.doOnError { release(a, ExitCase.Error(it)) }
.doOnCancel { release(a, ExitCase.Cancelled) }
.doOnComplete { release(a, ExitCase.Completed) }
```

I've added a test to ensure the expected behaviour in cancellation.

JorgeCastilloPrz replied:
I was mistaken about how the callbacks work, sorry. I'm refactored it to use the 3 different callbacks depending on the case now:

```
.doOnError { release(a, ExitCase.Error(it)) }
.doOnCancel { release(a, ExitCase.Cancelled) }
.doOnComplete { release(a, ExitCase.Completed) }
```

I've added a test to ensure the expected behaviour in cancellation.

JorgeCastilloPrz replied:
I was mistaken about how the callbacks work, sorry. I'm refactored it to use the 3 different callbacks depending on the case now:

```
.doOnError { release(a, ExitCase.Error(it)) }
.doOnCancel { release(a, ExitCase.Cancelled) }
.doOnComplete { release(a, ExitCase.Completed) }
```

I've added a test to ensure the expected behaviour in cancellation.

JorgeCastilloPrz replied:
I was mistaken about how the callbacks work, sorry. I'm refactored it to use the 3 different callbacks depending on the case now:

```
.doOnError { release(a, ExitCase.Error(it)) }
.doOnCancel { release(a, ExitCase.Cancelled) }
.doOnComplete { release(a, ExitCase.Completed) }
```

I've added a test to ensure the expected behaviour in cancellation.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/typeclasses/Bracket.kt:
```diff
@@ -0,0 +1,83 @@
+   * @param release is the action that's supposed to release the allocated resource after `use` is done, irregardless
+   * of its exit condition.
+   */
+  fun <A, B> Kind<F, A>.bracket(use: (A) -> Kind<F, B>, release: (A) -> Kind<F, Unit>): Kind<F, B> =
```
For readability purposes, I'd move the use to the last parameter. That way we have the execution always right-aligned.

`bla.bracket({ a -> callNetwork()}) { a.cancel() }`

vs 

`bla.bracket({ a.cancel() }) { a -> callNetwork()}`

JorgeCastilloPrz replied:
Turns out I kept the same ordering found in cats mostly because of the bracket semantics: You take the resource, use it and then release it. But your point is also good though.

pakoito replied:
I prefer the release/use style, for the reasons above. We can ask in the channel.

pakoito replied:
Reaffirming myself on the function order here :D

JorgeCastilloPrz replied:
Yeye, don't think I forgot, that'll be done before merging.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/CancelUtils.kt:
```diff
@@ -0,0 +1,59 @@
+import arrow.effects.IO
+import arrow.effects.IOFrame
+import arrow.effects.fix
+import com.sun.tools.javac.util.ListBuffer
```
We shouldn't depend on sun tools.

JorgeCastilloPrz replied:
Removed on later pushes, sorry, I had some code pending to push

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/CancelUtils.kt:
```diff
@@ -0,0 +1,59 @@
+/**
+ * utilities for dealing with cancelable thunks.
+ */
+internal object CancelUtils {
```
Yes, check this branch for them: https://github.com/arrow-kt/arrow/pull/984

JorgeCastilloPrz commented:
Hey @pakoito I fixed / answered some of your comments, also pushed new changes needed to fix all previously broken stuff from `IOBracket`. Tests are passing now. I think I've lost some of your comments after rebasing or something. Could you review again this new version? It should be more solid than the one you reviewed.

Still I have some concerns about the cancellation implementation. Here I'm relying on the new `IOConnection` which is ported over from cats and is used from a couple of places to achieve cancelation, recovering and a couple more things. Looks like it works good, but it probably doesn't fit well with what you shared on that cancellation related PR. I'm not sure if there's some overlapping or what. Any suggestions on how to move on and homogeneize in case it's needed?

I'd also love to get your thoughts on this @raulraja.

One love ❤️

pakoito commented on modules/core/arrow-instances-data/build.gradle:
```diff
@@ -3,6 +3,7 @@ dependencies {
     compile project(':arrow-annotations')
     compile project(':arrow-typeclasses')
     compile project(':arrow-data')
+    compile project(':arrow-effects')
```
This should be the other way around. Effects should have the MonadDefer instance of Kleisli.

JorgeCastilloPrz replied:
I'm probably missing something but, are you sure about this? The only instances I could find on `arrow-effects-instances` are the ones from `IO`, but just because it's an effects data type. Kleisli is not. That seems to be the approach until today.

The module does not contain the `MonadDefer` / `Bracket` / `Async` / `Effect` (or any other effects Type classes) instances for all the data types in Arrow.

nomisRev replied:
@JorgeCastilloPr This breaks the module system. If I include instances for `data` then I don't want to implicitly import `effects`.

What did you need `effects` for anyway?

pakoito replied:
Nevermind that, do a new module `arrow-data-instances-effects` instead.

JorgeCastilloPrz replied:
Alright. Did that as a compromise solution for now, mostly to unlock this PR. We definitely need to have that maintainers call asap to decide what to do with these instances, it doesn't look great though. Here's [the commit](https://github.com/arrow-kt/arrow/pull/1043/commits/9609ea0ce4945d4916d0ee8a107c5372b776bfdb).

I like the new approach with everything super segregated, but there are still some rough edges.

JorgeCastilloPrz replied:
Yep! removed now, sorry. I noticed some mins ago. The commit links got broken cause of rebases, you'd need to look at complete diff though.

pakoito commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt:
```diff
@@ -14,18 +14,18 @@ import kotlinx.coroutines.experimental.CommonPool
       Law("Sync bind: bindingInContext cancellation before flatMap") { SC.inContextCancellationBefore(EQ) },
       Law("Sync bind: bindingInContext cancellation after flatMap") { SC.inContextCancellationAfter(EQ) },
-      Law("Sync bind: bindingInContext throw equivalent to raiseError") { SC.inContextErrorThrow(EQERR) },
+      Law("Sync bind: bindingInContext throw equivalent to raiseError") { SC.inContextErrorThrow(EQ) },
```
Are you sure this needs to be changed?

JorgeCastilloPrz replied:
EQERR is an EQ instance for `Kind<F, Int>`, exactly the same than other param: `EQ`. If you look at callers, feels like we're passing the same instance twice everywhere:

![captura de pantalla 2018-11-10 a las 8 53 27](https://user-images.githubusercontent.com/6547526/48298950-23cb8c80-e4c6-11e8-8477-d1864597fa4f.png)

So I removed it since I didn't understand why we need two if they're just structural comparisons over the same `Kind<F, Int>`, so can't find a reason for them to be different. Please let me know if I'm wrong here. Actually I think I forgot to remove `EQERR` also in `AsyncLaws`, since it has the same issue.

pakoito replied:
Thinking about this, there's a potential issue where one of the tests returns an unexpected error, and our Eq accepts it because it's made to compare errors. Same for error cases, they may succeed and the Eq treats them as correct.

This is probably an issue with the callers, more than it is with the laws.

JorgeCastilloPrz replied:
Alright, I can see the problem now. I've reverted it to be as it was til today, since it doesn't look like a boyscout we wanna do here, given the highlighted caveats of the approach. Still, it might be something to address in the maintainer's meeting.

pakoito commented on modules/effects/arrow-effects-reactor/src/main/kotlin/arrow/effects/FluxK.kt:
```diff
@@ -80,7 +94,7 @@ data class FluxK<A>(val flux: Flux<A>) : FluxKOf<A>, FluxKKindedJ<A> {
     fun <A> runAsync(fa: Proc<A>): FluxK<A> =
       Flux.create { emitter: FluxSink<A> ->
-        fa { either: Either<Throwable, A> ->
+        fa(IOConnection()) { either: Either<Throwable, A> ->
```
Can't we register the `IOConnection` to work in sync with `emitter`? AFAIK they represent the same semantics.

JorgeCastilloPrz replied:
`IOConnection` is not exposed through Async anymore, mimicking cats.

pakoito commented on modules/effects/arrow-effects-reactor/src/main/kotlin/arrow/effects/MonoK.kt:
```diff
@@ -58,7 +68,7 @@ data class MonoK<A>(val mono: Mono<A>) : MonoKOf<A>, MonoKKindedJ<A> {
     fun <A> async(fa: Proc<A>): MonoK<A> =
       Mono.create { emitter: MonoSink<A> ->
-        fa { either: Either<Throwable, A> ->
+        fa(IOConnection()) { either: Either<Throwable, A> ->
```
Same comment about `emitter` and `IOConnection`

JorgeCastilloPrz replied:
`IOConnection` is not exposed through Async anymore, mimicking cats.

pakoito commented on modules/effects/arrow-effects-rx2/src/main/kotlin/arrow/effects/FlowableK.kt:
```diff
@@ -28,6 +34,14 @@ data class FlowableK<A>(val flowable: Flowable<A>) : FlowableKOf<A>, FlowableKKi
+  fun <B> bracketCase(use: (A) -> FlowableKOf<B>, release: (A, ExitCase<Throwable>) -> FlowableKOf<Unit>): FlowableK<B> =
+    flatMap { a ->
+      use(a).fix().flowable
+        .doOnNext { release(a, ExitCase.Completed) }
```
doOnComplete

pakoito commented on modules/effects/arrow-effects-reactor/src/main/kotlin/arrow/effects/MonoK.kt:
```diff
@@ -27,6 +29,14 @@ data class MonoK<A>(val mono: Mono<A>) : MonoKOf<A>, MonoKKindedJ<A> {
+  fun <B> bracketCase(use: (A) -> MonoKOf<B>, release: (A, ExitCase<Throwable>) -> MonoKOf<Unit>): MonoK<B> =
+    flatMap { a ->
+      use(a).fix().mono
+        .doOnNext { release(a, ExitCase.Completed) }
```
Check if onNext == onComplete for Mono

JorgeCastilloPrz replied:
Actually it's not, the correct one was `onSuccess()`, which checks the `Mono` has completed successfully. Fixed it and added a test to cover cancellation.

pakoito commented on modules/effects/arrow-effects-rx2/src/main/kotlin/arrow/effects/ObservableK.kt:
```diff
@@ -27,6 +33,14 @@ data class ObservableK<A>(val observable: Observable<A>) : ObservableKOf<A>, Obs
+  fun <B> bracketCase(use: (A) -> ObservableKOf<B>, release: (A, ExitCase<Throwable>) -> ObservableKOf<Unit>): ObservableK<B> =
+    flatMap { a ->
+      use(a).fix().observable
+        .doOnNext { release(a, ExitCase.Completed) }
```
doOnComplete

JorgeCastilloPrz replied:
Fixed. Following [this](https://github.com/ReactiveX/RxJava/issues/4456) and [this](https://github.com/ReactiveX/RxJava/pull/4458) looks like `Observable.onDispose()` would be the equivalent to `Flowable.onCancel()`, so I'm notifying the `Cancelled` ExitCase when the observer is disposed.

Added a test to cover the cancellation.

pakoito commented on modules/effects/arrow-effects-rx2/src/main/kotlin/arrow/effects/SingleK.kt:
```diff
@@ -58,7 +68,7 @@ data class SingleK<A>(val single: Single<A>) : SingleKOf<A>, SingleKKindedJ<A> {
     fun <A> async(fa: Proc<A>): SingleK<A> =
       Single.create { emitter: SingleEmitter<A> ->
-        fa { either: Either<Throwable, A> ->
+        fa(IOConnection()) { either: Either<Throwable, A> ->
```
Same about emitter

JorgeCastilloPrz replied:
`IOConnection` is not exposed through Async anymore, mimicking cats. Further explanations in a comment in the end.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IO.kt:
```diff
@@ -90,7 +100,7 @@ sealed class IO<out A> : IOOf<A> {
 
   fun runAsyncCancellable(onCancel: OnCancel = Silent, cb: (Either<Throwable, A>) -> IOOf<Unit>): IO<Disposable> =
-    IO.async { ccb ->
+    IO.async { conn, ccb ->
```
It seems like this is the point where my cancellation should be replaced with yours.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IOInstances.kt:
```diff
@@ -67,7 +81,22 @@ interface IOMonadErrorInstance : IOMonadInstance, MonadError<ForIO, Throwable> {
-interface IOMonadDeferInstance : IOMonadErrorInstance, MonadDefer<ForIO> {
+interface IOBracketInstance : IOMonadErrorInstance, Bracket<ForIO, Throwable> {
+  override fun <A, B> Kind<ForIO, A>.bracketCase(use: (A) -> Kind<ForIO, B>, release: (A, ExitCase<Throwable>) -> Kind<ForIO, Unit>): IO<B> =
+    this@bracketCase.fix().bracketCase({ a -> use(a).fix() }, { a, e -> release(a, e).fix() })
```
This shouldn't be needed here

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IOParallel.kt:
```diff
@@ -18,7 +18,7 @@ fun <A, B, C> IO.Companion.parallelMapN(ctx: CoroutineContext, ioA: IO<A>, ioB:
   }
 
-  return IO.async { cb: (Either<Throwable, C>) -> Unit ->
+  return IO.async { _, cb: (Either<Throwable, C>) -> Unit ->
```
Cancellation for Parallels needs to be setup here.

JorgeCastilloPrz replied:
`IOConnection` is not exposed through Async anymore, mimicking cats. Further explanations in a comment in the end.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt:
```diff
@@ -0,0 +1,112 @@
+  /**
+   * Default [[IOConnection]] implementation.
+   */
+  private class Impl : IOConnection() {
```
I'd call it a different name, like `DefaultIOConnection()`.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt:
```diff
@@ -0,0 +1,112 @@
+   * Guaranteed idempotency - calling it multiple times should have the same side-effect as calling it only
+   * once. Implementations of this method should also be thread-safe.
+   */
+  abstract fun cancel(): CancelToken<ForIO>
```
All the `CancelToken` in the PR are fixed to `ForIO`. Doesn't this imply that there should be other fixes for other datatypes?

JorgeCastilloPrz replied:
Yep. We need to support Cancellation over any `F` but that's part of [ConcurrentEffect](https://github.com/typelevel/cats-effect/blob/76d354153926b30217cc1aed9d1b181c7ea20d32/site/src/main/tut/typeclasses/concurrent-effect.md) and [Concurrent](https://github.com/typelevel/cats-effect/blob/77931fd2fc6d7d2891e321cd59689a47b2e13a56/site/src/main/tut/typeclasses/concurrent.md). That will be done when we work on those types completely (we just have `ConcurrentEffect` on very early steps for now).

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOPlatform.kt:
```diff
@@ -0,0 +1,21 @@
+package arrow.effects.internal
+
+internal object IOPlatform {
```
We don't really need this object because IO will never work outside the JVM. Merge it in another file if you want.

nomisRev replied:
> never work outside the JVM

We never plan to support IO on js?

JorgeCastilloPrz replied:
These functions are used from a couple of places so not sure where to place them. I've renamed the file to `ErrorUtils` for now, following the same style than `CancelUtils`.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/ParallelUtils.kt:
```diff
@@ -14,7 +14,7 @@ import kotlin.coroutines.experimental.suspendCoroutine
 internal fun <F, A, B, C> Effect<F>.parMap2(ctx: CoroutineContext, ioA: Kind<F, A>, ioB: Kind<F, B>, f: (A, B) -> C,
   /* start is used because this has to start inside the coroutine. Using Future won't work */
-                                            start: (Kind<F, Unit>) -> Unit): Proc<C> = { cc ->
+                                            start: (Kind<F, Unit>) -> Unit): Proc<C> = { _, cc ->
```
Cancellation needs to be setup here.

JorgeCastilloPrz replied:
`IOConnection` is not exposed through Async anymore, mimicking cats. Further explanations in a comment in the end.

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt:
```diff
@@ -0,0 +1,112 @@
+  /**
+   * Pushes a cancelable reference on the stack, to be popped or canceled later in FIFO order.
+   */
+  abstract fun push(token: CancelToken<ForIO>): Unit
```
I understand the semantics behind this function name. But, for the user's sake, I'd replace it with something more like `push` -> `register` and `pop` -> `unregisterLatest`

pakoito replied:
More inline with `setDisposable` and `setCancellable` here http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/ObservableEmitter.html#setDisposable-io.reactivex.disposables.Disposable-

JorgeCastilloPrz replied:
Actually this is not a stack, but a queue. It's **FIFO**, so `unregisterLatest` wouldn't be accurate either. Ideas?

nomisRev commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/typeclasses/Bracket.kt:
```diff
@@ -0,0 +1,83 @@
+import arrow.effects.typeclasses.ExitCase.Error
+import arrow.typeclasses.MonadError
+
+sealed class ExitCase<out E> {
```
Should this be moved to `arrow.effects` like cats?

JorgeCastilloPrz replied:
[It comes along with `Bracket` into the same file in `cats-effect`](https://github.com/typelevel/cats-effect/blob/dbbbf3c30bd8b52d31bd95b3084aad1054cb483a/core/shared/src/main/scala/cats/effect/Bracket.scala#L146), so did the same here.

raulraja commented:
This is ready to for a rebase from master. The Extensions PR is in already

JorgeCastilloPrz commented:
Ok GitHub is trolling me and can't comment this into an outdated comment. @pakoito:

I was mistaken about how the callbacks work for `FluxK`, sorry. I've refactored it now to use the 3 different callbacks depending on the case:

```
.doOnError { release(a, ExitCase.Error(it)) }
.doOnCancel { release(a, ExitCase.Cancelled) }
.doOnComplete { release(a, ExitCase.Completed) }
```

Also added a test to ensure the expected behaviour in cancellation.

pakoito commented on modules/effects/arrow-effects-reactor/src/main/kotlin/arrow/effects/FluxK.kt:
```diff
@@ -35,9 +35,10 @@ data class FluxK<A>(val flux: Flux<A>) : FluxKOf<A>, FluxKKindedJ<A> {
+      use(a).value()
         .doOnError { release(a, ExitCase.Error(it)) }
+        .doOnCancel { release(a, ExitCase.Cancelled) }
+        .doOnComplete { release(a, ExitCase.Completed) }
```
👍

pakoito commented on modules/effects/arrow-effects-reactor/src/test/kotlin/arrow/effects/FluxKTests.kt:
```diff
@@ -1,58 +1,52 @@
 
   init {
-
+/*
```
Removed for testing?

JorgeCastilloPrz replied:
Oops sorry, commented this for reducing noise amount when testing something else. Uncommented now 👍

JorgeCastilloPrz commented on modules/effects/arrow-data-instances-effects/gradle.properties:
```diff
@@ -0,0 +1,4 @@
+# Maven publishing configuration
+POM_NAME=Arrow Type Classes Effects Instances for data types in Data
```
:D the fact that I can't describe this properly with human words could highlight the "smell" on the approach. Anyways don't wanna get into the rabbit hole more than what I already did, so keeping it for now.

pakoito replied:
"Instances of arrow-effects for the datatypes in arrow-data"

JorgeCastilloPrz replied:
Done

JorgeCastilloPrz commented:
After some cats digging, I made my mind around how `IOConnection` is used.

Turns out I initially added `IOConnection` as a requirement to `Proc` (`Async` callback), and that was incorrect. Following cats implementation, `IOConnection` is not part of the processing callback but a separate requirement by `IO.Async`. So it’s an ad hoc cancellation solution for `IO`. I’ve fixed it to follow the same constraints it has in typelevel/cats now.

That removes the need for passing in the `IOConnection` for each async() implementation on each data type. That was wrong since data types should not be tied to IO in any way.

About achieving a generalized way to control cancellation **across any data types supporting `Async`**, sounds totally legit, but I couldn’t find any relation between that concept and `IOConnection` in cats. I think that’s part of `Concurrent` and `ConcurrentEffect` work since those are indeed using the concept of `CancelToken<F>`, but they use it differently (not through `IOConnection`). Therefore that’d be a bit too out of scope of this PR and more in the line of #1090.

(IOConnection was required on this one because it’s part of the `IOBracket` implementation).

Said that, could I get a third review pass on everything @pakoito @nomisRev ? Thanks guys ❤️ .

JorgeCastilloPrz commented on modules/core/arrow-data/build.gradle:
```diff
@@ -12,6 +12,7 @@ dependencies {
     testCompile project(':arrow-test')
     testCompile project(':arrow-free')
     testCompile project(':arrow-instances-free')
+    testCompile project(':arrow-data-instances-effects')
```
This is required since tests are all on this module but this particular instance is in the newly added module.

JorgeCastilloPrz commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/ForwardCancelable.kt:
```diff
@@ -0,0 +1,89 @@
+/**
+ * A placeholder for a [[CancelToken]] that will be set at a later time, the equivalent of a
+ * `Deferred[IO, CancelToken]`. Used in the implementation of `bracket`, see [[IOBracket]].
+ */
```
fixed

JorgeCastilloPrz commented on modules/effects/arrow-effects-instances/build.gradle:
```diff
@@ -1,5 +1,7 @@
 dependencies {
     compile project(':arrow-effects')
+    compile project(':arrow-data')
+    compile project(':arrow-instances-data')
```
That's removed already, sorry, missed it!

JorgeCastilloPrz commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/ErrorUtils.kt:
```diff
@@ -0,0 +1,21 @@
+   * triggered, is not acceptable.
+   *
+   * On top of the JVM this function uses Throwable#addSuppressed, available since Java 7. On top of JavaScript the
+   * function would return a CompositeException.
```
Looks like it does. This is docs just copied over from cats because seemed to be important to highlight.

JorgeCastilloPrz commented on modules/dagger/arrow-dagger/src/test/kotlin/arrow/dagger/instances/tests/instances.kt:
```diff
@@ -24,7 +63,7 @@ class LocalOptionTTInstances : OptionTInstances<ForOption>()
 
 @Module
-class LocalKleisliInstances : KleisliInstances<ForOption, Unit>()
+class LocalKleisliInstances : KleisliInstances<ForOption, Unit, Unit>()
```
This is not needed anymore. KleisliInstances contained the bracket instance before and therefore it required the error type, and that one does not belong to this dagger instances module.

pakoito commented on modules/core/arrow-data/src/test/kotlin/arrow/data/KleisliTest.kt:
```diff
@@ -17,19 +28,34 @@ import io.kotlintest.KTestJUnitRunner
 import io.kotlintest.matchers.shouldBe
 import org.junit.runner.RunWith
 
+@Suppress("TestFunctionName")
```
What is this annotation for?

JorgeCastilloPrz replied:
Otherwise you get the lint warning about the test names

![captura de pantalla 2018-11-11 a las 18 02 21](https://user-images.githubusercontent.com/6547526/48315867-fbce4d00-e5db-11e8-953f-eaa790480d53.png)

pakoito commented on modules/core/arrow-instances-data/src/main/kotlin/arrow/instances/kleisli.kt:
```diff
@@ -33,69 +51,54 @@ interface KleisliContravariant<F, D> : Contravariant<Conested<Kind<ForKleisli, F
 interface KleisliApplicativeInstance<F, D> : Applicative<KleisliPartialOf<F, D>>, KleisliFunctorInstance<F, D> {
 
-  fun AF(): Applicative<F>
+  override fun FFF(): Applicative<F>
```
Can you do this? We've a few of these through the codebase, and we reasigned them instead, like:

override fun FFF() = AFF()

fun AFF(): Applicative<F>

JorgeCastilloPrz replied:
Added #1104 to work on that separately. Ideally we'd have the complete instance hierarchy for all the data types following this approach which reduces the amount of boilerplate quite a bit.

pakoito commented on modules/effects/arrow-effects-instances/src/main/kotlin/arrow/effects/instances/io.kt:
```diff
@@ -4,9 +4,22 @@ import arrow.Kind
+import arrow.typeclasses.Monad
+import arrow.typeclasses.MonadError
+import arrow.typeclasses.Monoid
+import arrow.typeclasses.Semigroup
```
I suspect your IDE is configured not to collapse these after seeing 5+ of them.

JorgeCastilloPrz replied:
True story 👍

JorgeCastilloPrz commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IORunLoop.kt:
```diff
@@ -133,31 +133,27 @@ internal object IORunLoop {
-          val rcb = RestartCallback(cb as Callback, null)
-          rcb.prepare(bFirst, bRest)
-          register(conn, rcb)
+          loop(currentIO, conn, cb as Callback, null, bFirst, bRest)
```
why was this so different than cats? I'm not 100% sure of this change, but this is what you can find in cats for this function

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt:
```diff
@@ -3,8 +3,15 @@ package arrow.effects.internal
+import arrow.effects.typeclasses.Disposable
 import java.util.concurrent.atomic.AtomicReference
 
+fun IOConnection.toDisposable(): Disposable = object : Disposable {
```
Disposable is an alias for `() -> Unit`, so `= { cancel() }` should be enough

pakoito commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IORunLoop.kt:
```diff
@@ -239,6 +247,20 @@ internal object IORunLoop {
           bFirst = currentIO as BindF
           currentIO = currentIO.source
         }
+        is IO.ContextSwitch -> {
```
EDIT: I don't really know what this does..

JorgeCastilloPrz replied:
It's 2 different reifications of IO.

* `IO.ContinueOn` switches coroutine context.
* `IO.ContextSwitch` represents an `IOConnection` switch. It tries to modify the current connection using it's `modify` function with type `(IOConnection) -> IOConnection`. If it's udpated successfully, it updates the connection in the `RestartCallback` so next time it's invoked it'll be updated. `ContextSwitch` is also able to restore a failed connection using it's `restore` function which relies on the `RestoreContext`.

pakoito commented on modules/dagger/arrow-dagger/src/test/kotlin/arrow/dagger/instances/tests/instances.kt:
```diff
@@ -1,9 +1,48 @@
+import arrow.typeclasses.Order
+import arrow.typeclasses.Semigroup
+import arrow.typeclasses.SemigroupK
+import arrow.typeclasses.Traverse
```
Configure your imports to fold after a 5+ of them. This becomes unmaintainable after a while.

JorgeCastilloPrz replied:
Done! Fixed all over the place now.

pakoito commented on modules/effects/arrow-effects-instances/src/test/kotlin/arrow/effects/IOTest.kt:
```diff
@@ -243,7 +243,7 @@ class IOTest : UnitSpec() {
 
     "unsafeRunTimed times out with None result" {
-      val never = IO.async<Int> { }
+      val never = IO.async().never<Unit>().fix()
```
*squint* I wrote it, but it's still ugly. Maybe it's being generated now and it's just an import?

JorgeCastilloPrz replied:
No idea

JorgeCastilloPrz commented on modules/aql/arrow-query-language/src/test/kotlin/arrow/aql/tests/IOExample.kt:
```diff
@@ -39,7 +39,7 @@ class RemoteDataSource : DataSource {
 
   override fun allTasksByUser(user: User): IO<List<Task>> =
-    IO.async { cb ->
+    IO.async { _, cb ->
```
I think this one should be ignored. As in: no matter what connection the new async gets passed, it'll reuse the one passed in for the current `loop()` call.

JorgeCastilloPrz commented on modules/effects/arrow-effects/src/main/kotlin/arrow/effects/IORunLoop.kt:
```diff
@@ -233,8 +233,8 @@ internal object IORunLoop {
 
-          currentIO = IO.async { TODO, cc ->
-            loop(localCont, conn, cc.asyncCallback(currentCC), null, null, null)
+          currentIO = IO.async { _, callback ->
```
I think this one should be ignored. As in: no matter what connection the new async gets passed, it'll reuse the one passed in for the current `loop()` call. Can't see a reason to use a modified one or anything here, since there's no new cancellation reasons. Does it make sense to you @pakoito ?

pakoito replied:
Add a comment there, just in case, to explain your reasoning.

raulraja commented:
@JorgeCastilloPrz Is this good to go?, feel free to merge when conflicts are resolved if you and @nomisRev are happy with it

JorgeCastilloPrz commented:
It still requires some connection related work over IOParallel

JorgeCastilloPrz commented:
Hey @pakoito, I've been trying to understand how to wire the new IOConnection for cancellation on IOParallel utilities, but I've not been able to. I fear I don't have a deep scope on how that works and I'd say there's no IOConnection cancellation wired in cats for this. Since you probably were the one working on that it would be faster if you give some help with that. Could you plssss? 🙏

JorgeCastilloPrz commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/AsyncLaws.kt:
```diff
@@ -13,8 +13,14 @@ import io.kotlintest.properties.forAll
 object AsyncLaws {
-  fun <F> laws(AC: Async<F>, EQ: Eq<Kind<F, Int>>, EQ_EITHER: Eq<Kind<F, Either<Throwable, Int>>>, EQERR: Eq<Kind<F, Int>> = EQ): List<Law> =
-    MonadDeferLaws.laws(AC, EQERR, EQ_EITHER, EQ) + listOf(
+
```
Looks like around 50% of the Laws have the blank line at the beginning and the other 50% don't. Added it to all of them now for coherence.

JorgeCastilloPrz replied:
Fixed

JorgeCastilloPrz commented on modules/core/arrow-instances-data/src/main/kotlin/arrow/instances/kleisli.kt:
```diff
@@ -33,69 +33,54 @@ interface KleisliContravariant<F, D> : Contravariant<Conested<Kind<ForKleisli, F
-  fun MF(): Monad<F>
-
-  override fun AF(): Applicative<F> = MF()
+  override fun FFF(): Monad<F>
```
Nope. I mean, it was working good but @raulraja informed that it was not okay to do that. Reverting it now.

JorgeCastilloPrz commented on modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt:
```diff
@@ -13,8 +13,14 @@ import kotlinx.coroutines.Dispatchers
 object MonadDeferLaws {
-  fun <F> laws(SC: MonadDefer<F>, EQ: Eq<Kind<F, Int>>, EQ_EITHER: Eq<Kind<F, Either<Throwable, Int>>>, EQERR: Eq<Kind<F, Int>> = EQ): List<Law> =
-    MonadErrorLaws.laws(SC, EQERR, EQ_EITHER, EQ) + listOf(
+
```
fixed

JorgeCastilloPrz commented:
Added a comment on #1091 to work on `IOParallel` cancellation through `IOConnection` whenever `Parallel` is coded.
